services/contacts/js/app.js

```diff
@@ -0,0 +1,165 @@
+          continue;
+        }
+
+        if (typeof mozContact[key] !== 'function') {
+          cloned[key] = mozContact[key];
+        }
+      }
+      return cloned;
+    }
+
+    function _updateContact(realObj, fakeObj) {
+      for (var key in realObj) {
+        if (typeof realObj[key] === 'function') {
+          continue;
+        }
+
+        if (realObj[key] instanceof Date) {
+          if (+realObj[key] !== +fakeObj[key]) {
+            realObj[key] = fakeObj[key];
+          }
+          continue;
+        }
+
+        if (typeof realObj[key] === 'object') {
+          realObj[key] = _updateContact(realObj[key], fakeObj[key]);
+          continue;
+        }
+
+        if (realObj[key] !== fakeObj[key]) {
+          realObj[key] = fakeObj[key];
+        }
+      }
+
+      return realObj;
+    }
+
+    function sendError(error) {
+      channel.postMessage({
+        remotePortId: remotePortId,
+        data: { id : request.id, error: error.name }}
+      );
+    }
+
+    function sendResult(result) {
+      channel.postMessage({
+        remotePortId: remotePortId,
+        data: { id : request.id, result: result }}
+      );
+    }
+
+    function listenerTemplate(evt) {
+      channel.postMessage({
+        remotePortId: remotePortId,
+        data: {
+          id: request.id,
+          event: {
+            contactID: evt.contactID,
+            reason: evt.reason
+          }
+        }
+      });
+    }
+
+    if (requestOp.operation === 'oncontactchange') {
+      _contacts.oncontactchange = listenerTemplate;
+    } else if (requestOp.operation === 'getAll') {
+      var cursor = _contacts.getAll.apply(_contacts, requestOp.params);
+      var contacts = [];
+
+      cursor.onsuccess = () => {
+        var contact = cursor.result;
+        // 'cursor.done' flag should be activated when the last file is
+        // reached. However, it seems that the flag is only is enabled in 
+        // the next iteration so we've always got an undefined file
+        if (typeof contact !== 'undefined') {
+          contacts.push(_cloneContact(contact));
+        }
+
+        if (!cursor.done) {
+          cursor.continue();
+        } else {
+          // Send message
+          sendResult(contacts);
+        }
+      };
+
+      cursor.onerror = () => {
+        sendError(cursor.error);
+      };
+    } else if (requestOp.operation === 'save') {
+      var fakeContact = requestOp.params[0];
+      console.info(fakeContact.id, fakeContact.givenName);
+      var filter = {
+        filterBy: ['id'],
+        filterValue: fakeContact.id,
+        filterOp: 'equals'
+      };
+      var realContact;
+
+      _contacts.find(filter).then(result => {
+        if (!result.length) {
+          sendError({ name: 'No contacts found' });
+          return;
+        }
+        // Need to update realContact fields
+        var updatedContact = _updateContact(result[0], fakeContact);
+        _contacts.save(updatedContact).then(() => {
```

Comment:
.then(sendResult) 
or 
.then(() => sendResult(null))
(if null and undefined are not the same for that function)